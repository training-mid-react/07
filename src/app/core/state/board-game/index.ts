export * from './boardGameActions';
export * from './boardGameCases';
export * from './boardGameInitialState';
export * from './actions';
