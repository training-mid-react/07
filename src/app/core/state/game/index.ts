import { COLS, PLAYER1, ROWS } from "../../constants/constants";
import { CELL, IGameState } from "../../interfaces/game.interface";
import { IPlayer } from "../../interfaces/player.interface";
import { gameActions } from "./actions";

export const gameInitialState: IGameState = {
    grid: Array(ROWS).fill(Array(COLS).fill(null)),
    playerActive: PLAYER1,
    winner: null,
    isDraw: false,
};

export const gameCases = {
    [gameActions.CHANGE_PLAYER]: (state: IGameState, payload: IPlayer) => {
        return {
            ...state,
            playerActive: payload,
        };
    },
    [gameActions.SET_WINNER]: (state: IGameState, payload: IPlayer) => {
        return {
            ...state,
            winner: payload,
        };
    },
    [gameActions.CHANGE_GRID]: (state: IGameState, payload: CELL[][]) => {
        return {
            ...state,
            grid: payload,
        };
    },
    [gameActions.SET_DRAW]: (state: IGameState, payload: boolean) => {
        return {
            ...state,
            isDraw: payload,
        };
    },
    [gameActions.RESET_GAME]: (state: IGameState, payload: IGameState) => {
        return {
            ...payload,
        };
    },
}