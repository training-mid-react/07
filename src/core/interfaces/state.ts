import { GameState } from ".";

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface IState extends GameState {
}

export interface IContext {
  state: IState;
  dispatch: React.Dispatch<IAction>;
}

export interface IAction {
  type: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  payload: any;
}