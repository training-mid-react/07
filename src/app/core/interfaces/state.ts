import { CHANGE_PLAYER, DROP_PIECE, RESET_GAME, SET_WINNER } from "../state/cards/actions";

export interface IGameState {
  board: number[][];
  currentPlayer: number;
  winner: number | null;
  draw: boolean;
}

export type GameAction =
  | { type: typeof DROP_PIECE; payload: number }
  | { type: typeof CHANGE_PLAYER; payload: number }
  | { type: typeof SET_WINNER; payload: number }
  | { type: typeof RESET_GAME; payload: number }

export interface IAppContextProps {
  state: IGameState;
  dispatch: React.Dispatch<any>;
}