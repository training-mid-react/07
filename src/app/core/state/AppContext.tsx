import { createContext, useContext, useReducer } from 'react';
import { initialState, reducer } from './reducer';
import { IContext, IState } from '@core/interfaces';

export const AppContext = createContext<IContext>({
    state: {} as IState,
    dispatch: () => {},
});

export const AppProvider = ({ children }: { children: React.ReactNode }) => {
    const [state, dispatch] = useReducer(reducer, initialState);

    return (
        <AppContext.Provider value={{ state, dispatch }}>
            {children}
        </AppContext.Provider>
    );
};

export const useAppContext = () => useContext(AppContext);
