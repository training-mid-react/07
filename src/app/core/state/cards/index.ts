import { IGameState } from "../../interfaces/state";
import { handleDropPiece } from "../../utils/gameUtils";
import { gameInitialState } from "../reducers";
import { CHANGE_PLAYER, DROP_PIECE, RESET_GAME, SET_DRAW, SET_WINNER } from "./actions";

export const gameCases = {
    [DROP_PIECE]: (state: IGameState, payload: number) => {
        return {
            ...state,
            board: handleDropPiece(state.board, payload, state.currentPlayer),
        };
    },
    [CHANGE_PLAYER]: (state: IGameState) => {
        return {
            ...state,
            currentPlayer: state.currentPlayer === 1 ? 2 : 1, 
        };
    },
    [SET_WINNER]: (state: IGameState, payload: number) => {
        return {
            ...state,
            winner: payload, 
        };
    },
    [SET_DRAW]: (state: IGameState) => {
      return {
          ...state,
          draw: true, 
      };
  },
    [RESET_GAME]: () => {
        return {
            ...gameInitialState, 
        };
    },
};

