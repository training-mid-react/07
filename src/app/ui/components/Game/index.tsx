import { HeaderBoard } from '../HeaderBoard';
import { Board } from '../Board';
import { FC } from 'react';
import { IState } from '@core/interfaces';

export const Game: FC<{
    state: IState;
    onClickResetButton: () => void;
    handleClick: (col: number) => void;
    fallingCells: string[];
}> = ({ state, onClickResetButton, handleClick, fallingCells }) => (
    <>
        <HeaderBoard state={state} onClickResetButton={onClickResetButton} />
        <Board
            handleClick={handleClick}
            fallingCells={fallingCells}
            board={state.board}
        />
    </>
);
