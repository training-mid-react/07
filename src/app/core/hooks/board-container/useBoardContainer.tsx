import { ROWS } from '@core/constants';
import { AppContext } from '@core/state/AppContext';
import {
    resetState,
    setPlayeres,
    setWinner,
    updateBoard,
} from '@core/state/board-game/actions';
import { checkWinner } from '@core/utils';
import { useContext, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { IBoardContainerProps } from '../../interfaces/game/boardContainerProps.interface';

export const useBoardContainer = (): IBoardContainerProps => {
    const { state, dispatch } = useContext(AppContext);
    const navigate = useNavigate();
    const winner = state.winner;
    const board = state.board;
    const players = state.players;
    const [fallingCells, setFallingCells] = useState<string[]>([]);

    const onClickResetButton = () => {
        dispatch(resetState());
        navigate('/');
    };

    const handleClick = (col: number) => {
        if (winner) return;

        for (let row = ROWS - 1; row >= 0; row--) {
            if (board && !board[row][col]) {
                const newBoard = [...board];
                const currentPlayer = players?.find((p) => p.isCurrentPlayer);
                newBoard[row][col] = currentPlayer?.color ?? null;

                setFallingCells((prev) => [...prev, `${row}-${col}`]);
                dispatch(updateBoard(newBoard));

                const result = checkWinner({
                    board: newBoard,
                    row,
                    col,
                    players,
                });

                if (result) dispatch(setWinner(result));

                const [player1, player2] = players ?? [];
                dispatch(
                    setPlayeres([
                        {
                            ...player1,
                            isCurrentPlayer: !player1.isCurrentPlayer,
                        },
                        {
                            ...player2,
                            isCurrentPlayer: !player2.isCurrentPlayer,
                        },
                    ])
                );
                return;
            }
        }
    };

    return { state, fallingCells, onClickResetButton, handleClick };
};
