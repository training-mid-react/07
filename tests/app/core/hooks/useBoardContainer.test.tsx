import { vi, describe, test, expect } from 'vitest';
import { IState } from '../../../../src/app/core/interfaces/state.interface';
import { render, fireEvent, screen } from '@testing-library/react';
import { AppContext } from '../../../../src/app/core/state/AppContext';
import React from 'react';
import { useBoardContainer } from '../../../../src/app/core/hooks/board-container/useBoardContainer';
import { Game } from '../../../../src/app/ui/components/Game/index';
import '@testing-library/jest-dom';

vi.mock('react-router-dom', () => ({
    useNavigate: vi.fn().mockImplementation(() => vi.fn()),
}));

const initialState: IState = {
    winner: '',
    board: Array(6).fill(Array(7).fill(null)),
    players: [
        { id: 1, name: 'Jugador 1', color: 'blue', isCurrentPlayer: true },
        { id: 2, name: 'Jugador 2', color: 'red', isCurrentPlayer: false },
    ],
};
const mockDispatch = vi.fn(() => {
    return {
        navigate: vi.fn(),
    };
});

describe('useBoardContainer hook', () => {
    const renderWithContext = ({ children }: { children: React.ReactNode }) => {
        return render(
            <AppContext.Provider
                value={{ state: initialState, dispatch: mockDispatch }}
            >
                {children}
            </AppContext.Provider>
        );
    };

    const TestComponent = () => {
        const props = useBoardContainer();

        return <Game {...props} />;
    };

    test('Match snapshoot', () => {
        const { asFragment } = renderWithContext({
            children: <TestComponent />,
        });

        expect(asFragment()).toMatchSnapshot();
    });

    test('should return initial state and functions', async () => {
        renderWithContext({ children: <TestComponent /> });

        expect(mockDispatch).not.toHaveBeenCalled();
    });

    test('should reset the state when onClickResetButton is called', async () => {
        renderWithContext({ children: <TestComponent /> });
        const restartButton = screen.getByText('Reiniciar juego');
        fireEvent.click(restartButton);

        expect(mockDispatch).toHaveBeenCalled();
    });
});
