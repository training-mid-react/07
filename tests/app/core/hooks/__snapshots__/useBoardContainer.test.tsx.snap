// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useBoardContainer hook > Match snapshoot 1`] = `
<DocumentFragment>
  <div
    class="board-layout__header"
  >
    <section>
      <h3>
        Jugadores:
      </h3>
      <ul>
        <li
          style="color: blue;"
        >
          Jugador 1
        </li>
        <li
          style="color: red;"
        >
          Jugador 2
        </li>
      </ul>
    </section>
    <h1>
      Jugador 1 es tu turno
    </h1>
    <button>
      Reiniciar juego
    </button>
  </div>
  <div
    class="board"
  >
    <div
      class="board__column"
    >
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
    </div>
    <div
      class="board__column"
    >
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
    </div>
    <div
      class="board__column"
    >
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
    </div>
    <div
      class="board__column"
    >
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
    </div>
    <div
      class="board__column"
    >
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
    </div>
    <div
      class="board__column"
    >
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
      <div
        class="board__square"
      >
        <div
          class="board__square__content "
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
